Rework ForEachStmt Desugaring

elsakeirouz opened the pull request:
This PR aims to lift the desugaring of ForEachStmt (when dealing with Sequences) from SILGen to the AST, and simultaneously delaying the synthesis of makeIterator and next calls until after typechecking.

The motivation for this is support for non-Copyable & non-Escapable sequences with non-Copyable elements in ForEachStmt. In order to know which variant of makeIterator and next calls we need, we must determine whether a Sequence is copyable, asynchronous, or neither. The compiler currently relies on the function choice to infer the type of the elements. However, to determine whether a Sequence is copyable, it must be typed.
For that reason, instead of matching the pattern type with the unwrapped result type of next(), we're now matching it with Sequence. This allows us to delay the decision and synthesis of those function calls until later.

Additionally, in order to avoid having to translate the rest of the complex desugar logic for the borrowing ForEachStmt in SILGen, we chose to represent it directly at the AST level, which SILGen will then use to emit the corresponding SIL code.

To make things simpler all around, we chose to start by reworking this pass for the regular/async ForEachStmts before handling borrowing statements.

We expect to be able to remove ConstraintKind::ValueWitness with this work.

elsakeirouz commented on test/Sema/editor_placeholders.swift:
```diff
@@ -33,7 +33,7 @@ f(<#T##String#>) // expected-error{{editor placeholder in source file}} expected
 `<#foo#>`(x:)<Int>; // expected-error {{cannot find '<#foo#>(x:)' in scope}}
 
-for x in <#T#> { // expected-error{{editor placeholder in source file}} expected-error{{for-in loop requires '()' to conform to 'Sequence'}}
+for x in <#T#> { // expected-error{{value of type '_' has no member 'makeIterator()'}} expected-error{{editor placeholder in source file}} expected-error{{for-in loop requires '()' to conform to 'Sequence'}}
```
Oh sure, will look into that. Thanks!

hamishknight reviewed (commented):
Some initial comments from a quick look though, this is looking really good!

elsakeirouz commented on lib/Sema/Constraint.cpp:
```diff
@@ -1157,4 +1162,4 @@ void Constraint::setPreparedOverload(PreparedOverload *preparedOverload) {
           preparedOverload->wasForDiagnostics()));
 
   Overload.Prepared = preparedOverload;
-}
```
Undo this change

hamishknight committed:
[CS] Remove ValueWitness constraint

The specialized type-checking for `for` loops was the only client of
this, it's now unused.

hamishknight committed:
[CS] Remove some now-unnecessary `for` loop logic

We no longer need to track the `ForEachStmtInfo` in the
`SyntacticElementTarget`, and we can remove the special diagnostic
logic for `next` and `makeIterator` since those are type-checked
separately now.

elsakeirouz committed:
[Sema] desugar ForEachStmt at AST level

elsakeirouz committed:
[AST] Introduce opaque AST nodes

elsakeirouz committed:
[CS] introduce ForEachElement constraint

hamishknight commented on include/swift/AST/Expr.h:
```diff
@@ -6724,6 +6724,27 @@ class MacroExpansionExpr final : public Expr,
+
+  Expr *getOriginalExpr() const { return OriginalExpr; }
+  SourceLoc getStartLoc() const { return OriginalExpr->getStartLoc(); }
+  SourceLoc getEndLoc() const { return OriginalExpr->getEndLoc(); }
```
TODO: Consistency with pattern?

hamishknight commented on include/swift/AST/Pattern.h:
```diff
@@ -876,6 +897,8 @@ inline Pattern *Pattern::getSemanticsProvidingPattern() {
   if (auto *vp = dyn_cast<BindingPattern>(this))
     return vp->getSubPattern()->getSemanticsProvidingPattern();
+  if (auto *op = dyn_cast<OpaquePattern>(this))
+    return op->getSubPattern()->getSemanticsProvidingPattern();
```
Ideally we wouldn't do this by default, can you check to see which clients require it?

jckarter commented on include/swift/Sema/Constraint.h:
```diff
@@ -213,6 +208,8 @@ enum class ConstraintKind : char {
   /// The first type is a l-value type whose object type is the second type.
   LValueObject,
+  /// The first type of the sequence. The second type is the element type.
+  ForEachElement,
```
Can this be formulated in terms of existing constraint kinds, by binding the `$Element` type variable to `$Sequence.Element` encoded as `DependentMemberType($Sequence, Sequence.Element)`?

hamishknight replied:
Unfortunately not since we need to handle existential opening. Ideally we'd completely replace DependentMemberType in the constraint system with a proper constraint (since DMTs don't properly record fixes when they fail), in which case we could unify the implementation by having existential opening be an extra flag

jckarter commented on test/AutoDiff/compiler_crashers_fixed/issue-62608-conflicting-debug-info.swift:
```diff
@@ -37,7 +37,7 @@ public func pullback<T>(
 
 // CHECK-LABEL: sil private @$s4main19testUpdateByCallingyyKF8fOfArrayL_5arraySdSaySdG_tFySdzcfU_TJpSUpSr :
-// CHECK: alloc_stack $Double, var, name "derivative of 'element' in scope at {{.*}} (scope #3)"
+// CHECK: alloc_stack $Double, var, name "derivative of 'element' in scope at {{.*}} (scope #4)"
```
If the `(scope #)` is irrelevant to the test outcome, it might be good to remove it from the `CHECK:` line altogether so that the test doesn't need to change again in the future.

elsakeirouz replied:
I'm not sure it is irrelevant, moreso it makes sense that it's in a lower scope now that it's nested in a BraceStmt + WhileStmt instead of just ForEachStmt.

hamishknight commented on lib/AST/Pattern.cpp:
```diff
@@ -311,6 +321,8 @@ void Pattern::forEachNode(llvm::function_ref<void(Pattern*)> f) {
   case PatternKind::Binding:
     return cast<BindingPattern>(this)->getSubPattern()->forEachNode(f);
+  case PatternKind::Opaque:
+    return cast<OpaquePattern>(this)->getSubPattern()->forEachNode(f);
```
I don't think we should be recursing into opaque patterns here, similar to the ASTWalker. We'll probably want to handle this case specifically in `isRefutablePattern` since that should recurse

hamishknight commented on lib/Sema/CSGen.cpp:
```diff
@@ -2721,7 +2721,12 @@ namespace {
-        auto underlyingType = ogPattern->getType();
+        // Recursively handle the sub-pattern to ensure it gets properly
+        // registered in the constraint system
+        auto underlyingType = getTypeForPattern(
```
Did you figure out why we needed this?

hamishknight commented on lib/Sema/TypeCheckPattern.cpp:
```diff
@@ -1150,6 +1156,33 @@ Pattern *TypeChecker::coercePatternToType(
+
+    sub = coercePatternToType(
+        pattern.forSubPattern(sub, /*retainTopLevel=*/false), type, subOptions,
+        tryRewritePattern);
```
Did we figure out why this is needed? We shouldn't need to recurse into the underlying pattern

hamishknight commented on include/swift/AST/Stmt.h:
```diff
@@ -1023,51 +1023,30 @@ class ForEachStmt : public LabeledStmt {
-  OpaqueValueExpr *elementExpr = nullptr;
-  Expr *convertElementExpr = nullptr;
+  BraceStmt *desugaredStmt = nullptr;
+  OpaqueStmt *opaqueStmt = nullptr;
```
For clarity, could we rename this to `opaqueBodyStmt`?

hamishknight commented on include/swift/AST/Stmt.h:
```diff
@@ -1096,20 +1075,32 @@ class ForEachStmt : public LabeledStmt {
+  void setDesugaredStmt(BraceStmt *newStmt) { desugaredStmt = newStmt; }
+
+  OpaqueStmt *getOpaqueStmt() { return opaqueStmt; }
+  void setOpaqueStmt(OpaqueStmt *newStmt) { opaqueStmt = newStmt; }
```
Same here

hamishknight commented on include/swift/AST/Stmt.h:
```diff
@@ -1096,20 +1075,32 @@ class ForEachStmt : public LabeledStmt {
+  WhileStmt *getContinueTarget() { return continueTarget; }
+
+  void setBreakTarget(WhileStmt *target) { breakTarget = target; }
+  WhileStmt *getBreakTarget() { return breakTarget; }
```
Could you add a doc comment to the getters for these?

hamishknight commented on include/swift/AST/TypeCheckerTypeIDZone.def:
```diff
@@ -674,3 +674,7 @@ SWIFT_REQUEST(TypeChecker, IsCustomAvailabilityDomainPermanentlyEnabled,
+
+SWIFT_REQUEST(TypeChecker, DesugarForEachStmtRequest,
+              Stmt*(const ForEachStmt*),
+              Cached, NoLocationInfo)
```
Nit, update to match the request

```suggestion
              BraceStmt *(ForEachStmt *),
              SeparatelyCached, NoLocationInfo)
```

slavapestov commented on include/swift/AST/Expr.h:
```diff
@@ -6724,6 +6724,27 @@ class MacroExpansionExpr final : public Expr,
+/// expr and where clause to avoid visiting it twice in the ASTWalker after
+/// having desugared the loop. This will only be processed in SILGen to emit
+/// the underlying expression.
+class OpaqueExpr final : public Expr {
```
How is this different from OpaqueValueExpr?

hamishknight commented on lib/AST/TypeCheckRequests.cpp:
```diff
@@ -2890,3 +2890,20 @@ void IsCustomAvailabilityDomainPermanentlyEnabled::cacheResult(
+std::optional<BraceStmt *> DesugarForEachStmtRequest::getCachedResult() const {
+  auto *fes = std::get<0>(getStorage());
+  auto *desugaredStmt = fes->getCachedDesugaredStmt();
+  if (!desugaredStmt) {
```
Sorry missed this before, but this isn't quite right since if the request itself returns `nullptr` we'll keep trying to evaluate it. I think we'll probably want to store an extra bit on the ForEachStmt to determine whether the request "is computed", you could use e.g `llvm::PointerIntPair<BraceStmt *, 1, bool>` for this

elsakeirouz replied:
Where should I be checking whether it's been computed to make sure we don't keep trying?

hamishknight commented on lib/Sema/TypeCheckEffects.cpp:
```diff
@@ -4632,13 +4640,22 @@ class CheckEffectsCoverage : public EffectsHandlingWalker<CheckEffectsCoverage>
+          // If we don't have a source location for this use, use the
+          // anchor instead.
+          if (unsafeUse.getLocation().isInvalid())
+            unsafeUse.replaceLocation(S->getForLoc());
```
Sorry I know I added this code but I just copy and pasted this bit, I think the comment needs tweaking 😄 

```suggestion
          // If we don't have a source location for this use, use the
          // location of the `for` instead.
          if (unsafeUse.getLocation().isInvalid())
            unsafeUse.replaceLocation(S->getForLoc());
```

hamishknight commented on lib/SILGen/SILGenStmt.cpp:
```diff
@@ -1453,170 +1453,10 @@ void StmtEmitter::visitForEachStmt(ForEachStmt *S) {
-  SGF.BreakContinueDestStack.pop_back();
+  SGF.OpaqueStmts[S->getOpaqueStmt()] = S->getBody();
+  auto *braceStmt = S->getDesugaredStmt();
+  if (braceStmt)
```
If the brace statement is null I think we ought to assert that we've at least emitted an error, e.g:

```cpp
  ASSERT(braceStmt || SGF.getASTContext().hadError());
```

elsakeirouz committed:
[AST] ForEachStmt: rename sequence getter/setter

elsakeirouz commented on lib/AST/Expr.cpp:
```diff
@@ -840,6 +841,7 @@ bool Expr::canAppendPostfixExpression(bool appendingPostfixOperator) const {
 
   case ExprKind::MacroExpansion:
   case ExprKind::CurrentContextIsolation:
+  case ExprKind::Opaque: /* This one should never occur */
```
It would only occur if type checking the desugared statement fails, indicating a bug in the compiler, and in which case a diagnostic might be more helpful than a crash.

elsakeirouz committed:
[DEBUGINFO] Tests the same things across architectures

elsakeirouz commented on test/DebugInfo/LoadableByAddress.swift:
The error was occurring on the macOS x86_64 test run but I was able to reproduce it locally. I don't really know why it required that specific architecture.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
### Performance (x86_64): -O

**Regression**                            | **OLD**    | **NEW**    | **DELTA** | **RATIO**
:---                                      | ---:       | ---:       | ---:      | ---:     
NSString.bridged.byteCount.ascii.macroman | 0.0        | 0.678      | +67800.0% | **0.00x (?)**
CxxStringConversion.cxx.to.swift          | 86.75      | 149.0      | +71.8%    | **0.58x (?)**
Data.hash.Medium                          | 32.569     | 40.317     | +23.8%    | **0.81x**
ObjectiveCBridgeStubDateAccess            | 154.625    | 184.692    | +19.4%    | **0.84x (?)**
InsertCharacterEndIndex                   | 85.423     | 96.273     | +12.7%    | **0.89x (?)**
Dictionary3OfObjects                      | 329.25     | 368.0      | +11.8%    | **0.89x**
Prims.NonStrongRef.UnownedSafe            | 525.5      | 568.333    | +8.2%     | **0.92x (?)**
Prims.NonStrongRef.UnownedSafe.Closure    | 525.5      | 568.0      | +8.1%     | **0.93x (?)**
PrefixArrayLazy                           | 16.977     | 18.313     | +7.9%     | **0.93x (?)**
&nbsp; | | | | 
**Improvement**                           | **OLD**    | **NEW**    | **DELTA** | **RATIO**
Dictionary4                               | 221.375    | 184.444    | -16.7%    | **1.20x (?)**
Monoids                                   | 20563442.0 | 17607031.0 | -14.4%    | **1.17x (?)**
StringAdder                               | 381.167    | 333.714    | -12.4%    | **1.14x (?)**
StringDistance.scalars.ascii              | 587.25     | 522.5      | -11.0%    | **1.12x (?)**
RomanNumbers2                             | 382.667    | 341.286    | -10.8%    | **1.12x (?)**
StringBuilder                             | 364.833    | 329.143    | -9.8%     | **1.11x (?)**
Dictionary4OfObjects                      | 237.667    | 215.333    | -9.4%     | **1.10x (?)**
RemoveWhereFilterStrings                  | 329.0      | 298.4      | -9.3%     | **1.10x (?)**
StringBuilderSmallReservingCapacity       | 371.167    | 337.0      | -9.2%     | **1.10x (?)**
String.replaceSubrange.Substring.Small    | 51.478     | 46.87      | -9.0%     | **1.10x (?)**
Array.removeAll.keepingCapacity.Int       | 0.102      | 0.093      | -8.7%     | **1.10x (?)**
Set.isSuperset.Seq.Empty.Int              | 54.25      | 49.519     | -8.7%     | **1.10x (?)**
Data.init.Sequence.513B.Count.I           | 35.054     | 32.097     | -8.4%     | **1.09x (?)**
String.replaceSubrange.RepChar            | 1533.0     | 1413.0     | -7.8%     | **1.08x (?)**
Data.init.Sequence.2049B.Count.I          | 11.047     | 10.25      | -7.2%     | **1.08x (?)**
EqualSubstringSubstring                   | 25.917     | 24.14      | -6.9%     | **1.07x (?)**

### Code size: -O

**Regression**                   | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                             | ---:  | ---:  | ---:  | ---:   
DictionaryKeysContains.o         | 7380  | 7476  | +1.3% | **0.99x**
DictTest3.o                      | 15335 | 15527 | +1.3% | **0.99x**
DictTest2.o                      | 9920  | 10032 | +1.1% | **0.99x**
DictionaryOfAnyHashableStrings.o | 7255  | 7335  | +1.1% | **0.99x**
&nbsp; | | | | 
**Improvement**                  | **OLD** | **NEW** | **DELTA** | **RATIO**
DictionaryCompactMapValues.o     | 14706 | 13746 | -6.5% | **1.07x**
ObserverForwarderStruct.o        | 2964  | 2932  | -1.1% | **1.01x**

### Performance (x86_64): -Osize

**Regression**                          | **OLD**  | **NEW**  | **DELTA** | **RATIO**  
:---                                    | ---:     | ---:     | ---:   | ---:       
MapReduceLazyCollectionShort            | 31.25    | 41.375   | +32.4% | **0.76x (?)**
MapReduceString                         | 30.24    | 39.08    | +29.2% | **0.77x (?)**
Data.hash.Medium                        | 30.435   | 37.692   | +23.8% | **0.81x (?)**
SequenceAlgosUnfoldSequence             | 21.807   | 26.218   | +20.2% | **0.83x (?)**
ObjectiveCBridgeStubDateAccess          | 130.714  | 152.533  | +16.7% | **0.86x (?)**
BucketSort                              | 143.125  | 163.923  | +14.5% | **0.87x (?)**
DictionaryLiteral                       | 2315.0   | 2582.857 | +11.6% | **0.90x (?)**
WordCountHistogramASCII                 | 2609.615 | 2848.077 | +9.1%  | **0.92x (?)**
&nbsp; | | | | 
**Improvement**                         | **OLD**  | **NEW**  | **DELTA** | **RATIO**  
NSString.bridged.byteCount.ascii.utf8   | 0.522    | 0.0      | -99.8% | **523.00x (?)**
MapReduceLazySequence                   | 87.714   | 44.034   | -49.8% | **1.99x**  
SequenceAlgosRange                      | 2831.667 | 1742.727 | -38.5% | **1.62x**  
EqualSubstringSubstring                 | 24.931   | 22.741   | -8.8%  | **1.10x (?)**
EqualSubstringSubstringGenericEquatable | 24.938   | 22.761   | -8.7%  | **1.10x (?)**

### Code size: -Osize

**Regression**           | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                     | ---:  | ---:  | ---:  | ---:   
DictionaryKeysContains.o | 6473  | 6542  | +1.1% | **0.99x**
DictTest3.o              | 10922 | 11038 | +1.1% | **0.99x**
DictionaryRemove.o       | 14805 | 14960 | +1.0% | **0.99x**

### Performance (x86_64): -Onone

**Regression**            | **OLD**  | **NEW**   | **DELTA** | **RATIO**
:---                      | ---:     | ---:      | ---:   | ---:     
PointerArithmetics        | 54475.0  | 67666.667 | +24.2% | **0.81x (?)**
Data.hash.Medium          | 34.727   | 42.167    | +21.4% | **0.82x (?)**
BitCount                  | 919.5    | 1109.0    | +20.6% | **0.83x (?)**
ExclusivityIndependent    | 9.013    | 10.8      | +19.8% | **0.83x (?)**
OpenClose                 | 77.85    | 92.6      | +18.9% | **0.84x (?)**
DataAccessBytesSmall      | 99.895   | 113.067   | +13.2% | **0.88x (?)**
MapReduceLazySequence     | 18435.0  | 20801.0   | +12.8% | **0.89x (?)**
StringWalk                | 3105.263 | 3496.667  | +12.6% | **0.89x (?)**
DataSubscriptSmall        | 74.05    | 83.111    | +12.2% | **0.89x (?)**
StrComplexWalk            | 4430.0   | 4947.5    | +11.7% | **0.90x (?)**
DataSubscriptMedium       | 56.625   | 63.158    | +11.5% | **0.90x (?)**
ArrayEqualShared          | 3909.0   | 4321.0    | +10.5% | **0.90x (?)**
DataCreateMedium          | 51375.0  | 56150.0   | +9.3%  | **0.91x (?)**
DataCreateSmall           | 8410.0   | 9055.0    | +7.7%  | **0.93x (?)**
&nbsp; | | | | 
**Improvement**           | **OLD**  | **NEW**   | **DELTA** | **RATIO**
DictionaryEqualShared     | 5585.0   | 3230.0    | -42.2% | **1.73x**
DictionaryEqualUnique     | 11031.0  | 7079.0    | -35.8% | **1.56x**
ArrayAppendGenericStructs | 1435.0   | 1100.0    | -23.3% | **1.30x (?)**
SetIsSubsetInt100         | 471.4    | 424.6     | -9.9%  | **1.11x (?)**
Set.isStrictSubset.Int50  | 275.625  | 254.429   | -7.7%  | **1.08x (?)**

### Code size: -swiftlibs

**Regression**            | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                      | ---:  | ---:  | ---:  | ---:   
libswiftObservation.dylib | 61440 | 65536 | +6.7% | **0.94x**


<details>
  <summary><strong>How to read the data</strong></summary>
The tables contain differences in performance which are larger than 8% and
differences in code size which are larger than 1%.

If you see any unexpected regressions, you should consider fixing the
regressions before you merge the PR.

**Noise**: Sometimes the performance results (not code size!) contain false
alarms. Unexpected regressions which are marked with '(?)' are probably noise.
If you see regressions which you cannot explain you can try to run the
benchmarks again. If regressions still show up, please consult with the
performance team (@eeckstein).
</details>

<details>
  <summary><strong>Hardware Overview</strong></summary>

      Model Name: Mac mini
      Model Identifier: Macmini8,1
      Processor Name: 6-Core Intel Core i7
      Processor Speed: 3.2 GHz
      Number of Processors: 1
      Total Number of Cores: 6
      L2 Cache (per Core): 256 KB
      L3 Cache: 12 MB
      Memory: 32 GB

</details>

elsakeirouz commented:
@swift-ci please benchmark

hamishknight commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -3430,3 +3439,284 @@ FuncDecl *TypeChecker::getForEachIteratorNextFunction(
+
+    sequenceProto = isAsync ? ctx.getProtocol(KnownProtocolKind::AsyncSequence)
+                            : ctx.getProtocol(KnownProtocolKind::Sequence);
+    seqConformanceRef = lookupConformance(seqType, sequenceProto);
```
I would add an assert to make sure this is either valid, or we have an existential sequence

hamishknight commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -3430,3 +3439,284 @@ FuncDecl *TypeChecker::getForEachIteratorNextFunction(
+        OptionalSomePattern::createImplicit(ctx, nextCallVarPattern);
+
+    auto PBI = ConditionalPatternBindingInfo::create(
+        stmt->getDeclContext()->getASTContext(), SourceLoc(), somePattern,
```
```suggestion
        ctx, SourceLoc(), somePattern,
```

hamishknight commented on include/swift/AST/Pattern.h:
```diff
@@ -892,13 +893,17 @@ class BindingPattern : public Pattern {
+  if (lookThroughOpaque) {
+    if (auto *op = dyn_cast<OpaquePattern>(this))
+      return op->getSubPattern()->getSemanticsProvidingPattern();
+  }
```
These all need to pass `lookThroughOpaque` to the recursive `getSemanticsProvidingPattern` calls

hamishknight commented:
Benchmark results excluding those not present in the previous results (so there should hopefully be some more signal here):

### Performance (x86_64): -O

**Regression**                           | **OLD**    | **NEW**    | **DELTA** | **RATIO**
:---                                     | ---:       | ---:       | ---:   | ---:     
ObjectiveCBridgeStubDateAccess           | 200.5      | 234.0      | +16.7% | **0.86x (?)**
&nbsp; | | | | 
**Improvement**                          | **OLD**    | **NEW**    | **DELTA** | **RATIO**
Dictionary4                              | 216.143    | 175.625    | -18.7% | **1.23x (?)**
Monoids                                  | 25679518.0 | 21742617.0 | -15.3% | **1.18x**
Dictionary4OfObjects                     | 239.6      | 203.167    | -15.2% | **1.18x (?)**
RemoveWhereFilterInts                    | 36.489     | 32.556     | -10.8% | **1.12x (?)**
StringDistance.scalars.ascii             | 568.25     | 522.75     | -8.0%  | **1.09x (?)**

### Code size: -O

**Regression**                   | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                             | ---:  | ---:  | ---:  | ---:   
DictionaryKeysContains.o         | 7380  | 7476  | +1.3% | **0.99x**
DictTest3.o                      | 15335 | 15527 | +1.3% | **0.99x**
DictTest2.o                      | 9920  | 10032 | +1.1% | **0.99x**
DictionaryOfAnyHashableStrings.o | 7255  | 7335  | +1.1% | **0.99x**
&nbsp; | | | | 
**Improvement**                  | **OLD** | **NEW** | **DELTA** | **RATIO**
DictionaryCompactMapValues.o     | 14706 | 13746 | -6.5% | **1.07x**
ObserverForwarderStruct.o        | 2964  | 2932  | -1.1% | **1.01x**

### Performance (x86_64): -Osize

**Regression**                 | **OLD**  | **NEW**  | **DELTA** | **RATIO**
:---                           | ---:     | ---:     | ---:   | ---:     
MapReduceLazyCollectionShort   | 31.267   | 41.4     | +32.4% | **0.76x (?)**
SequenceAlgosUnfoldSequence    | 21.848   | 26.218   | +20.0% | **0.83x**
ObjectiveCBridgeStubDateAccess | 130.765  | 152.533  | +16.6% | **0.86x (?)**
BucketSort                     | 143.25   | 164.143  | +14.6% | **0.87x (?)**
DictionaryLiteral              | 2280.0   | 2600.0   | +14.0% | **0.88x (?)**
WordCountHistogramASCII        | 2549.02  | 2775.51  | +8.9%  | **0.92x (?)**
&nbsp; | | | | 
**Improvement**                | **OLD**  | **NEW**  | **DELTA** | **RATIO**
MapReduceLazySequence          | 87.722   | 44.032   | -49.8% | **1.99x**
SequenceAlgosRange             | 2832.857 | 1742.727 | -38.5% | **1.63x**

### Code size: -Osize

**Regression**           | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                     | ---:  | ---:  | ---:  | ---:   
DictionaryKeysContains.o | 6473  | 6542  | +1.1% | **0.99x**
DictTest3.o              | 10922 | 11038 | +1.1% | **0.99x**
DictionaryRemove.o       | 14805 | 14960 | +1.0% | **0.99x**

### Performance (x86_64): -Onone

**Regression**                           | **OLD** | **NEW**   | **DELTA** | **RATIO**
:---                                     | ---:    | ---:      | ---:   | ---:     
PointerArithmetics                       | 54475.0 | 67566.667 | +24.0% | **0.81x (?)**
ExclusivityIndependent                   | 9.02    | 10.942    | +21.3% | **0.82x (?)**
BitCount                                 | 920.0   | 1109.5    | +20.6% | **0.83x (?)**
OpenClose                                | 78.591  | 92.647    | +17.9% | **0.85x (?)**
DataAccessBytesSmall                     | 99.905  | 113.067   | +13.2% | **0.88x (?)**
DataSubscriptSmall                       | 74.053  | 83.125    | +12.3% | **0.89x (?)**
DataSubscriptMedium                      | 56.636  | 63.182    | +11.6% | **0.90x (?)**
DataCreateMedium                         | 51425.0 | 56075.0   | +9.0%  | **0.92x (?)**
DataCreateSmall                          | 8410.0  | 9120.0    | +8.4%  | **0.92x (?)**
&nbsp; | | | | 
**Improvement**                          | **OLD** | **NEW**   | **DELTA** | **RATIO**
DictionaryEqualShared                    | 5572.0  | 3182.0    | -42.9% | **1.75x**
DictionaryEqualUnique                    | 11080.0 | 7095.0    | -36.0% | **1.56x**

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight committed:
[SILGen] Upgrade an assert

Make sure we always crash if we don't have a failure branch for a
refutable pattern, otherwise we can end up infinite looping.

hamishknight committed:
[AST] Add flag to isRefutablePattern to determine IsPattern behavior

IsPattern coercions aren't properly implemented since they rely on
doing a runtime cast, which can fail for things like function subtyping.
Make clients choose what behavior they want. Unfortunately we need to
preserve the current behavior in places since it's relied upon for
things like exhaustivity checking. We ought to properly implement
coercion handling, then we should be able to remove this.

hamishknight commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -3432,3 +3439,270 @@ FuncDecl *TypeChecker::getForEachIteratorNextFunction(
+    ConcreteDeclRef nextWitness;
+    if (stmt->getParsedSequence()->getType()->isExistentialType())
+      nextWitness = ConcreteDeclRef(nextFn);
+    else {
```
Let's also handle a null `nextFn` here, this doesn't really matter since it only happens if the stdlib is broken, but we handle it above and below so let's be consistent

```suggestion
    else if (nextFn) {
```

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please benchmark

hamishknight reviewed (approved):
LGTM modulo a couple of minor comments, this looks great!

hjyamauchi commented:
Hi @elsakeirouz it seems that bisecting for [this compiler assert failure](https://github.com/swiftlang/swift/issues/87290) converges to this change. The details including reproduction are there. Can you help debug it?

CC @hamishknight @kavon
